### ISSUE TITLE
hashFiles silently generating empty hash

### ISSUE THREAD

**Issue by rbtcollins** [j3t/jenkins-pipeline-cache-plugin#36]
I'm not sure whats going on yet, but I had inappropriate cache hits, and tracked it down to hashFiles giving the same signature when file content had changed.

Concretely, I write a number of strings to a file:
```
String hashStrings(List strings) {
    // need a java extension perhaps - the sandbox is blocking the
    // obvious .digest('SHA-1') method.
    // Manual salt to allow rekeying if needed
    input = ["serial-1"] + strings
    temp_path = pwd(tmp: true)
    hash_path = "${temp_path}/cargo-cache-key-vars"
    writeFile(file: hash_path, text: input.join('\n'))
    echo "hashing ${input.join('\n')}"
    hash = sh(script: "md5sum < ${hash_path}", returnStdout: true).split(' ')[0]
    hash_files_hash = hashFiles(hash_path)
    echo "${hash} ${hash_files_hash}"
    return hash
}
```

Which echos `e973e42cfa03aa8c5a826ecffb41431a` `d41d8cd98f00b204e9800998ecf8427e`

Now that hashFiles hash is a well known hash - a completely empty hash:
```
echo -n | md5sum
d41d8cd98f00b204e9800998ecf8427e
```

[This code](https://github.com/j3t/jenkins-pipeline-cache-plugin/blob/ad533b31a213db2db606eae37e7ee40cfe958690/src/main/java/io/jenkins/plugins/pipeline/cache/HashFilesStep.java#LL110C22-L111C62)
Seems fine IFF it is choosing the Path override.

So I'm not sure. Perhaps something to do with the k8s plugin :/.

**Comment by j3t** [comment]
What is the specific problem? Is there something I can do? Otherwise I would like close this ticket.

**Comment by rbtcollins** [comment]
hashFiles(path_to_existing_file) should either:
- return the hash of that one file
- error if it cannot read the file

I'm confused as to why its not returning the has of that one file, given I'm using hashFiles happily elsewhere. Possibly something to do with the pwd(tmp:true) use? 

If you want to close it - sure. My workaround is to not use hashFiles :- but others in future may be equally confused.

**Comment by j3t** [comment]
* `hashFiles` expects a pattern which may match multiple files
* matching files might not have any content

Keeping this in mind I would argue it is fine like it is, but I agree that this could be confusing.

Proposal:
* print a warning when the content is empty
* add a note in the docs (or a link to the [JavaDoc](https://github.com/j3t/jenkins-pipeline-cache-plugin/blob/main/src/main/java/io/jenkins/plugins/pipeline/cache/HashFilesStep.java#L32-L33))

**Comment by rbtcollins** [comment]
I agree that it should be possible to hash an empty set (e.g. we use that feature to hash both rust-toolchain and rust-toolchain.toml as two separate patterns. I think a warning would be great, since it would make it visible.

**Comment by j3t** [comment]
I thought about it and I guess it is more clear when `hashFiles` returns an empty (just an empty string, no md5 hash) in case no files were found or the content is empty.

**Comment by jan-kantert** [comment]
We struggled with this as well. What about this: By default raise an error but add an option to explicitly allow this.

Rational: Usually you do not want this to happen. It only manifests later during changes and then causes correctness issues (those are the worst). If this is intended behaviour you will notice instantly and can add the option to allow it.
